curs/curs-18-19-linking/README.md

```diff
@@ -0,0 +1,1297 @@
+Simbolul `_start`, adică entry pointul, adică adresa primei instrucțiuni executate, este echivalentul unei funcții care apelează funcția `main()`.
+Deci `main()` nu este entry pointul unui program, nu este prima funcție executată.
+O altă secvență de cod, numită tipic, la fel ca în acest caz, `_start`, este prima executată.
+Și aceasta apelează `main()`.
+
+Așadar, dându-se unul sau mai multe fișiere obiect, linkerul creează executabilul și stabilește entry pointul acestuia.
+Entry pointul are sens doar pentru fișiere executabile, **NU** pentru fișiere obiect.
+
+### Address binding
+
+Linkerul atașează fiecărui simbol din fișierul executabil rezultat o adresă.
+Aceste adrese vor fi folosite la încărcarea executabilului în memorie, la crearea procesului.
+O astfel de adresă este adresa entry pointului, adresa primei instrucțiuni ce va fi executată.
+
+Simbolurile din fișierele obiect nu au atribuite adrese.
+Adresele vor fi atribuite fiecărui simbol la linkare.
+Când investigăm un fișier obiect, "adresele" afișate sunt de fapt offseturile în cadrul secțiunilor.
+Comanda de mai jos afișează simbolurile din fișierul obiect `one.o`:
+```
+[..]/01-one-file$ nm one.o
+00000000 T increment
+00000013 T main
+00000000 D num_items
+```
+În secvența de mai sus ar părea că simbolul `increment` și simbolul `num_items` au acceași adresă.
+De fapt, simbolul `increment` este un simbol în secțiunea de cod (`.text`) în vreme ce simbolul `num_items` este un simbol în secțiunea de date (`.data`).
+"Adresele" afișate sunt offseturile în cadrul secțiunilor.
+Adică simbolul `increment` este la offsetul `0` în cadrul secțiunii de cod (adică este la începutul secțiunii).
+La fel, simbolul `num_items` este la offsetul `0` în cadrul secțiunii de date (adică, la fel, este la începutul secțiunii).
+Un alt simbol, `main` este la offsetul `0x13` în cadrul secțiuni de cod.
+
+În dezasamblarea codului fișierului obiect `one.o` observăm adresele (adică, de fapt, offseturile) pentru simbolurile `increment` și `main`:
+```
+[..]/01-one-file$ objdump -d -M intel one.o
+
+one.o:     file format elf32-i386
+
+
+Disassembly of section .text:
+
+0000000 <increment>:
+   0:	55                   	push   ebp
+   1:	89 e5                	mov    ebp,esp
+   3:	a1 00 00 00 00       	mov    eax,ds:0x0
+   8:	83 c0 01             	add    eax,0x1
+   b:	a3 00 00 00 00       	mov    ds:0x0,eax
+  10:	90                   	nop
+  11:	5d                   	pop    ebp
+  12:	c3                   	ret
+
+00000013 <main>:
+  13:	55                   	push   ebp
+  14:	89 e5                	mov    ebp,esp
+  16:	c7 05 00 00 00 00 05 	mov    DWORD PTR ds:0x0,0x5
+  1d:	00 00 00
+  20:	e8 fc ff ff ff       	call   21 <main+0xe>
+  25:	b8 00 00 00 00       	mov    eax,0x0
+  2a:	5d                   	pop    ebp
+  2b:	c3                   	ret
+```
+Adresele / offseturile sunt cele așteptate: `0` pentru `increment` și `0x13` pentru `main`<a href="#ds" id="refds"><sup>1</sup></a>.
```

Comment:
Ce e cu html-ul asta: <a href="#ds" id="refds"><sup>1</sup></a>.?

Reply:
E footnote. Nu există sintaxă dedicată în GitHub Markdown, dar poți face folosind cod HTML. Am preluat de [aici](https://stackoverflow.com/questions/25579868/how-to-add-footnotes-to-github-flavoured-markdown).